src/cunumeric/matrix/solve.cc

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/matrix/solve.h"
+#include "cunumeric/matrix/solve_template.inl"
+
+#include <cblas.h>
+#include <lapack.h>
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <>
+struct SolveImplBody<VariantKind::CPU, LegateTypeCode::FLOAT_LT> {
+  void operator()(int32_t m, int32_t n, int32_t nrhs, float* a, float* b)
+  {
+    const char trans = 'N';
+    int32_t info     = 0;
+
+    auto kind = CuNumeric::has_numamem ? Memory::Kind::SOCKET_MEM : Memory::Kind::SYSTEM_MEM;
+    auto ipiv = create_buffer<int32_t>(std::min(m, n), kind);
+
+    LAPACK_sgetrf(&m, &n, a, &m, ipiv.ptr(0), &info);
+    LAPACK_sgetrs(&trans, &n, &nrhs, a, &m, ipiv.ptr(0), b, &n, &info);
```

Comment:
I think we need to check the value of `info` between the two calls as well. The way this is written currently, the second call would override the first call's `info`.

Based on the documentation of [sgetrf](https://netlib.org/lapack/explore-html/d8/ddc/group__real_g_ecomputational_ga8d99c11b94db3d5eac75cac46a0f2e17.html), it is the first call whose output it is critical to check, otherwise the second call will throw division by zero:

```
          > 0:  if INFO = i, U(i,i) is exactly zero. The factorization
                has been completed, but the factor U is exactly
                singular, and division by zero will occur if it is used
                to solve a system of equations.
```

The other error conditions (`info < 0`, for both `getrf` and `getrs`) seem to be just that an argument has an illegal value. So I think overall we should always test if `info > 0` right after the first call (and throw a user-visible exception if so), and then `assert(info == 0)` in debug mode after both calls (since illegal arguments should have been caught by the frontend).

Reply:
Thanks for taking a closer look. https://github.com/nv-legate/cunumeric/pull/568/commits/838511747a0744ae3fd9c49dd111f9c7d580b18d should fix the issue in the GPU variant. For CPU variants, I actually rewrite them so that they call `_gesv` instead, as that's what NumPy is using (and it's simpler).